feat: implement albums feature with create, edit, delete, and image management UI

SiddharthJiyani opened the pull request:
## Description

This PR implements a complete Albums feature , including both backend API endpoints and a fully functional frontend UI. 

## Features Implemented

### Backend
- Improved Album CRUD operations (Create, Read, Update, Delete)
- Album locking/password protection
- Cover image management
- Image management within albums (add, remove, reorder)
- Proper error handling and validation

### Frontend
- Albums UI with album cards
- Create album dialog with name, description, and lock option
- Edit album dialog with ability to change lock status
- Delete album with confirmation dialog
- Password-protected access for locked albums
- Add/remove images from albums
- Set album cover image
- Dark/Light theme support with appropriate placeholder images

## Technical Details

### Architecture
- **State Management**: Redux with `albumsSlice.ts` for persistent state
- **API Integration**: `usePictoQuery` and `usePictoMutation` hooks
- **UI Components**: Standard ShadCN components (Dialog, Button, Input, Label, Switch, etc.)
- **User Feedback**: `infoDialogSlice.ts` for messages and `loaderSlice.ts` for loading states
- **Type Safety**: Full TypeScript coverage with proper interfaces

### Key Changes
1. Created Redux slice (`albumsSlice.ts`) with 11 actions for album state management
2. Created memoized selectors (`albumSelectors.ts`) using `createSelector`
3. Implemented API functions for all album operations (`albums.ts`)
4. Built comprehensive album components:
   - `AlbumCard.tsx` - Album display with lock icon
   - `CreateAlbumDialog.tsx` - New album creation
   - `EditAlbumDialog.tsx` - Album editing with lock management
   - `DeleteConfirmDialog.tsx` - Delete confirmation
   - `AlbumPasswordDialog.tsx` - Password entry for locked albums
   - `AddImagesToAlbumDialog.tsx` - Image management
5. Created album pages:
   - `Album.tsx` - Albums list view
   - `AlbumDetail.tsx` - Album detail view with images
6. Added proper error handling with detailed backend error messages

## Testing Checklist

- Create album (locked and unlocked)
- Edit album (change name, description, lock status)
- Delete album with confirmation
- Lock/unlock albums with password
- Add/remove images from album
- Set album cover image
- Auto-refresh UI after operations
- Error messages display properly
- Dark/Light theme support
- Responsive design on mobile and desktop



## Related Issues

- #546

## Tasks

- [x] Use standard ShadCN components to create the UI. (Refer to other components)
- [x] Store all the state information using Redux only. (Refer to similar files to know how to create a Redux slice and store the query results)
- [x] Use only `usePictoQuery` and `usePictoMutation` to handle API calling.
- [x] Utilize the standard `infoDialogSlice.ts` and `loaderSlice.ts` to display dialog messages and loaders, respectively.


https://github.com/user-attachments/assets/bb3fdb5e-6dc5-4372-ac4d-a8cc867a3e11



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Albums now support locking (password protection), cover images, per-album photo counts, and a new API to set an album's cover.

* **UI**
  * Full album UI: listing, detail/gallery, create, edit, delete, add-images, password prompt, album card, and empty-state components.

* **Behavior Changes**
  * Album visibility uses "is_locked"; album list no longer accepts a show-hidden parameter and payloads include is_locked, cover_image_path, and image_count.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SiddharthJiyani committed:
feat: implement albums feature with create, edit, delete, and image management UI

SiddharthJiyani committed:
test: update album tests to use is_locked instead of is_hidden

SiddharthJiyani committed:
refactor: address coderabbit review feedback for albums feature

SiddharthJiyani committed:
refactor code to address errors

SiddharthJiyani committed:
refactor: use useMutationFeedback for album mutations

SiddharthJiyani committed:
fix: remove unused imports from album dialogs

SiddharthJiyani committed:
fix: address code review feedback

SiddharthJiyani commented on frontend/src/pages/Album/Album.tsx:
```diff
@@ -1,5 +1,268 @@
+        }),
+      );
+    },
+  });
```
Done

SiddharthJiyani committed:
fix: address code review issues - htmlFor mismatch, stale closure, refetch error handling

SiddharthJiyani committed:
Merge branch 'main' into feat/albums-ui-and-backend

SiddharthJiyani committed:
resolve linting issues

rahulharpal1603 reviewed (commented):
<img width="515" height="188" alt="Image" src="https://github.com/user-attachments/assets/86598c2b-ee34-43a4-b474-a2de53803b16" />

The above error is occurring whenever I try to create an album or add images to the album.

SiddharthJiyani committed:
resolve album creation error

SiddharthJiyani commented:
> <img alt="Image" width="515" height="188" src="https://private-user-images.githubusercontent.com/51887323/553125982-86598c2b-ee34-43a4-b474-a2de53803b16.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE3NDEzOTAsIm5iZiI6MTc3MTc0MTA5MCwicGF0aCI6Ii81MTg4NzMyMy81NTMxMjU5ODItODY1OThjMmItZWUzNC00M2E0LWI0NzQtYTJkZTUzODAzYjE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjIyVDA2MTgxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmZjdlMDViMTE5OWQwOTE5NGJmMDY4ZDVhZjAyOTcwMDdkYjZlMmE2M2VkMjgyZmRkOWFkMTczOTY2MGNmYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bpOybX2jFicgoOgZ6SceFWXJAiadbZs5TjX7_lqtJB4">
> The above error is occurring whenever I try to create an album or add images to the album.

Done, please check.

rohan-pandeyy commented:
Hi @SiddharthJiyani, could we catch up regarding the merge of this PR over a meeting call in [AOSSIE's Discord Server Meeting Room](https://discord.com/channels/1022871757289422898/1345795402682859561)?

SiddharthJiyani committed:
Merge branch 'main' into feat/albums-ui-and-backend

SiddharthJiyani committed:
resolve linting issues

rohan-pandeyy reviewed (changes requested):
Let's follow the say UI for "Back" button as seen in AI tagging page when we open a particular face collection

Currently, the back button in Albums looks like this:

<img width="1919" height="598" alt="Image" src="https://github.com/user-attachments/assets/906b5024-3d82-4549-953b-0835cc4f58fd" />

So lets make this consistent with the existing Back button from face collection:

<img width="1919" height="461" alt="Image" src="https://github.com/user-attachments/assets/540eb835-fd13-4579-96da-3e51ae6684fb" />

---

Additionally, when you work on this change, please try and make the final UI somewhat like this:

<img width="1919" height="594" alt="Image" src="https://github.com/user-attachments/assets/8c3f8c0e-f3a8-442d-8131-d42097a8fbc6" />

rohan-pandeyy committed:
Merge upstream/main into feat/albums-ui-and-backend

Resolve conflicts in images.py (drop the orphaned memories banner comment;
upstream moved those queries to database/memories.py) and AppRoutes.tsx
(keep upstream's Videos/Memories/ModelManager/Search routes, add the
albums routes).

Update test_albums_db.py for this branch's is_hidden -> is_locked rename:
listing returns locked albums, the password gates opening.

rohan-pandeyy committed:
docs: regenerate openapi.json for the album API changes

rohan-pandeyy committed:
fix(albums): migrate databases that predate the is_locked rename

CREATE TABLE IF NOT EXISTS is a no-op on an existing albums table, so any
database created before this branch kept is_hidden and never gained
cover_image_path. Every album read then failed with "no such column:
is_locked", 500ing GET /albums/ for all upgrading users.

Add guarded ALTERs alongside the create, matching the pattern already used
for image_classes.score and image_embeddings.scored_signature.

rohan-pandeyy committed:
refactor(albums): use the shared GallerySortDropdown

The albums page predates GallerySortDropdown and grew its own inline
DropdownMenu. Generalise the shared component to take an options list
so it can serve both, defaulting to the gallery's best-match/date pair
so existing call sites are unchanged.
